load-gen/src/main/java/io/smartcat/data/loader/LoadGenerator.java

```diff
@@ -1,155 +1,105 @@
+    private final Worker<T> worker;
 
-    private final TokenBucket tokenBucket;
-
-    private final PulseGenerator pulseGenerator;
-
-    private Timer metricsTimer;
-
-    private LoadGenerator(Builder builder) {
-        this.dataCollector = new DataCollector(builder.dataSource, builder.targetRate * 10);
-        this.tokenBucket = new TokenBucket(0, builder.refillStrategy, builder.sleepStrategy);
-        this.pulseGenerator = new PulseGenerator(dataCollector, tokenBucket, builder.workTask, builder.collectMetrics);
-
-        this.metricsTimer = new Timer("pulse-generator-timer");
-        this.metricsTimer.scheduleAtFixedRate(new MetricsLogger(), 0, 1000);
-    }
+    private AtomicBoolean terminate = new AtomicBoolean(false);
 
     /**
-     * Start load generator.
+     * Constructs load generator with specified <code>dataSource</code>, <code>rateGenerator</code> and
+     * <code>worker</code>.
+     *
+     * @param dataSource Data source from which load generator polls data.
+     * @param rateGenerator Rate generator which generates rate based on time point.
+     * @param worker Worker which accepts data polled from <code>dataSource</code> at rate provided by
+     *            <code>rateGenerator</code>.
      */
-    public void start() {
-        pulseGenerator.start();
+    public LoadGenerator(DataSource<T> dataSource, RateGenerator rateGenerator, Worker<T> worker) {
+        this.dataSource = dataSource;
+        this.rateGenerator = rateGenerator;
+        this.worker = worker;
     }
 
     /**
-     * Metrics logger timer task.
+     * Starts load generator.
+     *
+     * @throws IllegalStateException When start is attempted after load generator was already stopped.
      */
-    private class MetricsLogger extends TimerTask {
-        @Override
-        public void run() {
-            LOGGER.debug("Generated {} pulses", pulseGenerator.getPulseCount());
+    public void start() {
+        checkState();
+        long beginning = System.nanoTime();
+        long previous = beginning;
+        infiniteWhile: while (true) {
+            if (terminate.get()) {
+                LOGGER.info("Load generator stopped.");
+                break;
+            }
+            long now = System.nanoTime();
+            long fromBeginning = now - beginning;
+            long elapsed = now - previous;
+            long rate = rateGenerator.getRate(fromBeginning);
+            long normalizedRate = normalizeRate(elapsed, rate);
+            if (normalizedRate > 0) {
+                previous += calculateConsumedTime(normalizedRate, rate);
+            }
+            for (int i = 0; i < normalizedRate; i++) {
+                if (!dataSource.hasNext(elapsed)) {
+                    LOGGER.info("Reached end of data source. Stopping load generator.");
+                    terminate.set(true);
+                    break infiniteWhile;
+                }
+                T data = dataSource.getNext(fromBeginning);
```

Comment:
`fromBeginning` is relative time from moment when load generator is started, this is value for which API contract is defined. Elapsed is just delta time passed from previous call. It is used to determine frequency of calls and based on that normalize rate for that period of time, since rate generator returns rate per second.